src/odemis/driver/autoscript_client.py

```diff
@@ -0,0 +1,2220 @@
+                else os.path.isdir(entity_path) and len([entity for entity in os.listdir(entity_path) if entity.endswith(".exe")]) == 1
+            ):
+                package = Package()
+                package.adapter = result.group(1)
+                package.bitness = result.group(2) + "bit"
+                package.version = result.group(3)
+                package.name = entity
+                package.path = entity_path
+                # If not a zip file, assign the path of the executable
+                if not is_zip:
+                    for entity in os.listdir(package.path):
+                        if entity.endswith(".exe"):
+                            package.path = os.path.join(package.path, entity)
+                            break
+                # Filter using adapter type and bitness
+                if bitness == package.bitness and adapter == package.adapter:
+                    version_package[package.version] = package
+        # Check if any version is newer than the current one
+        versions = list(version_package)
+        if versions:
+            # Find the latest version
+            latest_version = max(versions, key=lambda s: [int(u) for u in s.split(".")])
+            if current_version < latest_version:
+                logging.info("The latest version is {}.\n".format(latest_version))
+                latest_package = version_package[latest_version]
+                if "-dev" in latest_package.name:
+                    logging.warning(
+                        "{} is a development version.\n".format(latest_package.name)
+                    )
+                return latest_package
+    return None
+
+
+# information on compatible versions
+debug_connection_info = f"""PYRO 5: {pkg_resources.get_distribution('Pyro5').version},
+msgpack-numpy: {pkg_resources.get_distribution('msgpack-numpy').version},
+msgpack_numpy_file: {msgpack_numpy.__file__},
+msgpack version: {pkg_resources.get_distribution('msgpack').version}
+msgpack_file: {msgpack.__file__}
+
+This is likely a dependencies issue:
+make sure you have the following matching versions of msgpack and msgpack-numpy installed:
+msgpack==0.5.6, msgpack-numpy==0.4.4
+or
+msgpack==1.0.3 msgpack-numpy==0.4.8
+"""
+
+class FIBSEM(model.HwComponent):
+    """
+    Driver to communicate with autoscript software on TFS microscopes. autoscript is the software TFS uses to control their microscopes.
+    To use this driver the autoscript adapter developed by Delmic should be running on the TFS PC. Communication to the
+    Microscope server is done via Pyro5.
+    """
+
+    def __init__(self, name, role, children, address, port: str ='4242', daemon=None,
+                 **kwargs):
+        """
+        Parameters
+        ----------
+        address: str
+            server ip address for the microscope server (sim address is localhost)
+        port: str
+            server port of the Microscope server, default is '4242'
```

Comment:
I don't think you want to run them at the same time as they conflict with each other. So i think having them on the same port will cause an error if you try to run both, which I think is good rather than silently running both

Reply:
Okay, then it would probably be nice to have a check here and in the xt_client to make sure that you are connected to the correct adapter. Then you can have a bit nicer error reporting (in the block below) in case the xtadapter is running while you need the autoscript adapter or vice versa.

```
        try:
            self.server = Pyro5.api.Proxy(f"PYRO:Microscope@{address}:{port}")
            self.server._pyroTimeout = 30  # seconds
            self._swVersion = self.server.get_software_version()
            self._hwVersion = self.server.get_hardware_version()
            logging.debug(
                f"Successfully connected to autoscript server with software version {self._swVersion} and hardware"
                f"version {self._hwVersion}")
        except CommunicationError as err:
            raise HwError("Failed to connect to autoscript server '%s'. Check that the "
                          "uri is correct and autoscript server is"
                          " connected to the network. %s %s" % (address, err, debug_connection_info))
        except OSError as err:
            raise HwError("XT server reported error: %s." % (err,))
```